File: packages/api-client-py/client/host_client.py
import aiohttp
from urllib.parse import urlparse

class HostClient:
    def __init__(self, url: str) -> None:
        self.url = urlparse(url)

    async def get(self, url: str) -> str:
        async with aiohttp.ClientSession() as session:
            url = f'{self.url.geturl()}{url}'
            async with session.get(url) as resp:
                return await resp.text()
    
    async def post(self, url: str, headers: str = None, data=None, config=None) -> str:
        async with aiohttp.ClientSession() as session:
            url = f'{self.url.geturl()}{url}'
            async with session.post(url, headers=headers, data=data, params=config) as resp:
                return await resp.text()
    
    async def delete(self, url: str, headers: str = None) -> str:
        async with aiohttp.ClientSession() as session:
            url = f'{self.url.geturl()}{url}'
            async with session.delete(url, headers=headers) as resp:
                return await resp.text()
    
    async def get_data(self, seq_path: str) -> bytes:
        with open(seq_path, 'rb') as f:
            return f.read()
    
    async def send_stream(self, url: str, stream: str, options: dict):
        headers = {
            'content-type': options.get('type'),
            'x-end-stream': options.get('end')
        }
        config = { 'parse': options.get('parse_response') }
        return await self.post(url, headers, stream, config)
 
    async def list_sequences(self) -> str:
        url = f'/sequences'
        return await self.get(url)
    
    async def list_instances(self) -> str:
        url = f'/instances'
        return await self.get(url)

    async def send_sequence(self, file, app_config = None) -> str:
        url = f'/sequence'
        data = await self.get_data(file)
        return await self.post(url, data=data)

    async def get_sequence(self, id: str) -> str:
        url = f'/sequence/{id}'
        return await self.get(url)

    async def delete_sequence(self, id: str) -> str:
        url = f'/sequence/{id}'
        headers = {'Content-Type': 'application/json'}
        return await self.delete(url, headers=headers)

    async def get_instance_info(self, id: str) -> str:
        url = f'/instance/{id}'
        return await self.get(url)

    async def get_load_check(self) -> str:
        url = f'/load-check'
        return await self.get(url)

    async def get_version(self) -> str:
        url = f'/version'
        return await self.get(url)
   
    async def get_log_stream(self) -> str:
        url = f'/log'
        return await self.get(url)

    async def send_named_data(self, topic: str, stream: str, content_type: str, end: bool):
        data = {'type': content_type, 'end': end, 'parse_response': 'stream'}
        return await self.send_stream(f'topic/{topic}', stream, options=data)

    async def get_named_data(self, topic: str):
        return await self.get(f'topic/{topic}')

File: packages/types/src/api-client/host-client.ts
/// <reference lib="dom" />

import { Readable } from "stream";
import { LoadCheckStat } from "../load-check-stat";
import { KillMessageData } from "../messages";
import * as STHRestAPI from "../rest-api-sth";
import { PublicSTHConfiguration } from "../sth-configuration";
import { ClientUtils, HttpClient, SendStreamOptions } from "../client-utils";

export type InstanceInputStream = "stdin" | "input";
export type InstanceOutputStream = "stdout" | "stderr" | "output" | "log";
export type TopicOperationResponse = {
    operation: { id: string; status: string };
    result?: { topic?: { name: string; contentType: string; origin?: { type: "hub" | "space"; id: string } }; deleted?: boolean };
    error?: { code: string; message: string };
};

export declare class ClientProvider {
    client: HttpClient;
}

export declare class InstanceClient {
    get id(): string;

    static from(id: string, host: ClientProvider): InstanceClient;
    stop(timeout: number, canCallKeepalive: boolean): Promise<STHRestAPI.SendStopInstanceResponse>;
    kill(opts?: KillMessageData): Promise<STHRestAPI.SendKillInstanceResponse>;
    sendEvent(eventName: string, message: string): Promise<STHRestAPI.ControlMessageResponse>;
    getNextEvent(eventName: string): Promise<any>;
    getEvent(eventName: string): Promise<any>;
    getEventStream(eventName: string): Promise<import("stream").Readable>;
    getRPCClient(): HttpClient;
    getHealth(): Promise<STHRestAPI.GetHealthResponse>;
    getInfo(): Promise<STHRestAPI.GetInstanceResponse>;
    getStream(streamId: InstanceOutputStream): ReturnType<HttpClient["getStream"]>;
    sendStream(streamId: InstanceInputStream, stream: Parameters<HttpClient["sendStream"]>[1] | string, requestInit?: RequestInit, options?: SendStreamOptions): Promise<any>;
    sendInput(stream: Parameters<HttpClient["sendStream"]>[1] | string, requestInit?: RequestInit, options?: SendStreamOptions): Promise<any>;
    sendStdin(stream: Parameters<HttpClient["sendStream"]>[1] | string): Promise<any>;
}

export declare class SequenceClient {
    get id(): string;

    static from(id: string, host: ClientProvider): SequenceClient;

    constructor(apiBase: string, utils: ClientUtils | undefined);

    start(payload: STHRestAPI.StartSequencePayload): Promise<InstanceClient>;
    listInstances(): Promise<string[]>;
    getInstance(id: string, host: ClientProvider): Promise<InstanceClient>;
    getInfo(): Promise<STHRestAPI.GetSequenceResponse>;
    overwrite(stream: Readable): Promise<SequenceClient>;
}

export declare class HostClient implements ClientProvider {
    apiBase: string;
    client: ClientUtils;

    constructor(apiBase: string, utils: ClientUtils | undefined);

    listSequences(): Promise<STHRestAPI.GetSequencesResponse>;
    getSequenceId(sequenceName: string): Promise<string[]>;
    listInstances(): Promise<STHRestAPI.GetInstancesResponse>;
    listEntities(): Promise<STHRestAPI.GetEntitiesResponse>;
    getAuditStream(requestInit?: RequestInit): ReturnType<HttpClient["getStream"]>;
    getLogStream(requestInit?: RequestInit): ReturnType<HttpClient["getStream"]>;
    sendSequence(sequencePackage: Parameters<HttpClient["sendStream"]>[1], requestInit?: RequestInit, update?: boolean): Promise<SequenceClient>;
    /** Accepts a sequence ID or a stable sequence name. */
    getSequence(sequenceId: string): Promise<STHRestAPI.GetSequenceResponse>;
    deleteSequence(sequenceId: string, opts?: { force: boolean }): Promise<STHRestAPI.DeleteSequenceResponse>;
    /** Accepts an instance ID or a stable instance name. */
    getInstanceInfo(instanceId: string): Promise<STHRestAPI.GetInstanceResponse>;
    getLoadCheck(requestInit?: RequestInit): Promise<LoadCheckStat>;
    getVersion(): Promise<STHRestAPI.GetVersionResponse>;
    getStatus(): Promise<STHRestAPI.GetStatusResponse>;
    getConfig(): Promise<PublicSTHConfiguration>;
    sendNamedData<T>(topic: string, stream: Parameters<HttpClient["sendStream"]>[1], requestInit?: RequestInit, contentType?: string, end?: boolean): Promise<T>;
    sendTopic<T>(topic: string, stream: Parameters<HttpClient["sendStream"]>[1], requestInit?: RequestInit, contentType?: string, end?: boolean): Promise<T>;
    getNamedData(topic: string, requestInit?: RequestInit): ReturnType<HttpClient["getStream"]>;
    getTopic(topic: string, requestInit?: RequestInit, contentType?: string): ReturnType<HttpClient["getStream"]>;
    createTopic(topic: string, contentType: string): Promise<{ topicName: string }>;
    deleteTopic(topic: string): Promise<{ message: string }>;
    getTopics(): Promise<STHRestAPI.GetTopicsResponse>;
    getTopicsV2(): Promise<{ items: Array<{ name: string; contentType: string; origin?: { type: "hub" | "space"; id: string } }> }>;
    createTopicV2(topic: { name: string; contentType: string; origin?: { type: "hub" | "space"; id: string } }): Promise<TopicOperationResponse>;
    deleteTopicV2(topic: string): Promise<TopicOperationResponse>;
    getTopicV2(topic: string, requestInit?: RequestInit, contentType?: string): ReturnType<HttpClient["getStream"]>;
    sendTopicV2<T>(topic: string, stream: Parameters<HttpClient["sendStream"]>[1], requestInit?: RequestInit, contentType?: string, end?: boolean): Promise<T>;
    /** Reuses the existing client name; selector may be an instance ID or stable instance name. */
    getInstanceClient(id: string): InstanceClient;
    /** Reuses the existing client name; selector may be a sequence ID or stable sequence name. */
    getSequenceClient(id: string): SequenceClient;
    getManagerClient(apiBase: string, utils: ClientUtils | undefined): import("./manager-client").ManagerClient;
}
